# Data Cleaning Day Files From FirstRate Data
It would also be nice to have day data, and to make sure we can get long enough sequences, we'll keep the day data all the way back to 6/1/20, as that's when the market started recovering from the COVID-19 crash.

**Update**: Change of plans, since there seems to be missing data, and it would take unnecessary time to obtain/clean that data, we'll just stick to getting the day data since 10/1/20.

## Imports

In [7]:
import pandas as pd
import numpy as np
import yfinance as yf
from datetime import date, datetime
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.plotting import heatmap
import dask
import dask.dataframe as dd
from dask import delayed
from pyarrow.parquet import ParquetFile
import pyarrow as pa
from tqdm import tqdm

import tulipy as ti

import sklearn
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.decomposition import KernelPCA
from sklearn.decomposition import IncrementalPCA

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras import initializers
from tensorflow.keras.models import load_model
from tensorflow.keras.utils import register_keras_serializable
from tensorflow.keras.optimizers import SGD
import keras_tuner as kt
from keras_tuner import HyperParameters

import os
import sys
import warnings
import shutil

## Getting the list of tickers that we need 1-day data for
We have our 2,599 tickers that has complete, continuous, and cleaned 1-min data. Now let's make sure that data exists for the day data as well.

In [7]:
# Getting the paths can making sure the lengths seem appropriate (day data should have more since it's not filtered)
cleaned_1min_parq_path = '/Volumes/T7/Filtered-Cleaned-Parquet-ML'
stocks_1day_path = '/Volumes/T7/stocks1day-csv'

print("Number of 1-min filtered and cleaned tickers:", len(os.listdir(cleaned_1min_parq_path)))
print("Number of 1-day unfiltered tickers total:", len(os.listdir(stocks_1day_path)))

Number of 1-min filtered and cleaned tickers: 2599
Number of 1-day unfiltered tickers total: 7592


In [12]:
# Getting the list of the tickers in the cleaned_1min_parq_path folder:
ticker_list_filtered = []

for folder_name in os.listdir(cleaned_1min_parq_path):
    ticker = folder_name.split('_')[0]
    ticker_list_filtered.append(ticker)

len(ticker_list_filtered)

2599

## Checking to see if the FRD day data is worth using and cleaning
We want to see if the FRD 1-day data is appropriate for us to use. If it's too unclean, we're better off creating our own OHLCV values from the 1-min FRD data.

In [49]:
# Getting the total number of trading days a stock should have from 6/1/20-9/30/20 (using AAPL b/c it definitely has all the days)
AAPL_path = '/Volumes/T7/stocks1day-csv/AAPL_full_1day_UNADJUSTED.csv'
AAPL_df = pd.read_csv(AAPL_path)
AAPL_df['Datetime'] = pd.to_datetime(AAPL_df['Datetime'])

start_date = pd.to_datetime("2020-06-01")
end_date = pd.to_datetime("2020-10-01")
num_days_2020 = len(AAPL_df[(AAPL_df['Datetime'] >= start_date) & (AAPL_df['Datetime'] < end_date)])
print("Number of days each relevant ticker should have in 2020:", num_days_2020)

Number of days each relevant ticker should have in 2020: 86


In [50]:
# Getting list of tickers without all the day data
missing_days_tickers = []

# Making sure that each ticker in the 1-min folder 1) exists in the day folder and 2) has 86 days of data from 6/1/20-9/30/20
for ticker in tqdm(ticker_list_filtered):

    # Getting the ticker file name and path
    ticker_1day_file = ticker + "_full_1day_UNADJUSTED.csv"
    ticker_1day_path = os.path.join(stocks_1day_path, ticker_1day_file)

    # Reading in the csv and converting to Datetime column to proper datetime format
    df = pd.read_csv(ticker_1day_path)
    df['Datetime'] = pd.to_datetime(df['Datetime'])

    # Checking to see if the number of days is not equal to the expected number (86 days from 6/1/20-9/30/20)
    num_days_ticker = len(df[(df['Datetime'] >= start_date) & (df['Datetime'] < end_date)])    
    if num_days_ticker != num_days_2020:
        print(ticker, "only has", num_days_ticker, "days of day data.")
        missing_days_tickers.append(ticker)


  1%|▌                                       | 34/2599 [00:00<00:14, 171.04it/s]

ASLE only has 67 days of day data.
ASAN only has 0 days of day data.
ACCD only has 61 days of day data.


  4%|█▍                                      | 93/2599 [00:00<00:13, 189.93it/s]

AOUT only has 26 days of day data.
API only has 65 days of day data.
ANNX only has 46 days of day data.
ALIT only has 51 days of day data.


  6%|██▎                                    | 152/2599 [00:00<00:13, 182.00it/s]

AMWL only has 8 days of day data.
ALXO only has 51 days of day data.
ADV only has 84 days of day data.
CMPS only has 7 days of day data.


  8%|███▏                                   | 209/2599 [00:01<00:12, 184.85it/s]

AZEK only has 75 days of day data.
DOGZ only has 85 days of day data.
AEVA only has 74 days of day data.


 12%|████▋                                  | 314/2599 [00:01<00:14, 159.00it/s]

CHPT only has 78 days of day data.
CELU only has 83 days of day data.


 15%|█████▊                                 | 388/2599 [00:02<00:12, 176.20it/s]

BUR only has 74 days of day data.
IMTX only has 61 days of day data.


 16%|██████▍                                | 426/2599 [00:02<00:12, 179.22it/s]

CCI only has 85 days of day data.
DYN only has 8 days of day data.


 19%|███████▌                               | 505/2599 [00:02<00:11, 184.14it/s]

AVO only has 0 days of day data.
AVPT only has 0 days of day data.
BSY only has 4 days of day data.


 21%|████████▏                              | 544/2599 [00:03<00:10, 189.57it/s]

DNB only has 62 days of day data.
DM only has 85 days of day data.
FFIE only has 16 days of day data.
BNL only has 8 days of day data.


 28%|██████████▊                            | 720/2599 [00:04<00:09, 191.60it/s]

DADA only has 80 days of day data.
FROG only has 9 days of day data.


 29%|███████████▍                           | 765/2599 [00:04<00:08, 204.22it/s]

BIGC only has 38 days of day data.
HIMS only has 83 days of day data.
GBIO only has 75 days of day data.


 33%|████████████▊                          | 854/2599 [00:04<00:08, 209.06it/s]

GLSI only has 2 days of day data.
GOCO only has 53 days of day data.
LPRO only has 76 days of day data.


 35%|█████████████▊                         | 920/2599 [00:05<00:07, 212.67it/s]

FOUR only has 80 days of day data.
MP only has 53 days of day data.
ML only has 31 days of day data.


 37%|██████████████▍                        | 966/2599 [00:05<00:07, 214.89it/s]

GCMG only has 57 days of day data.
GDRX only has 4 days of day data.


 40%|███████████████                       | 1033/2599 [00:05<00:07, 209.51it/s]

ETWO only has 73 days of day data.
PTVE only has 8 days of day data.


 41%|███████████████▊                      | 1078/2599 [00:05<00:07, 216.86it/s]

CVAC only has 31 days of day data.
LUNG only has 0 days of day data.
LSPD only has 12 days of day data.
LMND only has 61 days of day data.
LI only has 42 days of day data.


 44%|████████████████▊                     | 1147/2599 [00:06<00:06, 215.76it/s]

JAMF only has 48 days of day data.
LEGN only has 80 days of day data.
LCID only has 7 days of day data.
LAC only has 0 days of day data.
LAZR only has 25 days of day data.
KYMR only has 26 days of day data.
BEPC only has 46 days of day data.


 46%|█████████████████▍                    | 1192/2599 [00:06<00:06, 214.45it/s]

MIR only has 27 days of day data.
CURI only has 74 days of day data.


 49%|██████████████████▍                   | 1263/2599 [00:06<00:05, 224.67it/s]

PRLD only has 2 days of day data.
PLBY only has 20 days of day data.
ORGN only has 16 days of day data.
ITOS only has 46 days of day data.


 51%|███████████████████▍                  | 1331/2599 [00:06<00:06, 209.29it/s]

PRCH only has 83 days of day data.
PMVP only has 2 days of day data.
PLRX only has 82 days of day data.
PLTR only has 0 days of day data.


 53%|████████████████████▏                 | 1377/2599 [00:07<00:05, 211.59it/s]

BEKE only has 32 days of day data.


 55%|████████████████████▊                 | 1421/2599 [00:07<00:05, 208.11it/s]

NNOX only has 26 days of day data.
OPEN only has 69 days of day data.


 58%|██████████████████████                | 1508/2599 [00:07<00:05, 208.56it/s]

NTST only has 32 days of day data.
NRIX only has 46 days of day data.
STR only has 79 days of day data.


 61%|███████████████████████               | 1574/2599 [00:08<00:04, 211.37it/s]

STEP only has 9 days of day data.
NCNO only has 54 days of day data.
MVST only has 83 days of day data.


 64%|████████████████████████▎             | 1664/2599 [00:08<00:04, 216.39it/s]

SNOW only has 9 days of day data.
EOSE only has 0 days of day data.


 66%|████████████████████████▉             | 1708/2599 [00:08<00:04, 209.20it/s]

RBOT only has 16 days of day data.


 68%|█████████████████████████▉            | 1776/2599 [00:09<00:03, 218.56it/s]

RNA only has 75 days of day data.
RPRX only has 73 days of day data.


 70%|██████████████████████████▋           | 1822/2599 [00:09<00:03, 224.32it/s]

SABA only has 0 days of day data.
RXT only has 38 days of day data.
XPEV only has 22 days of day data.
CR only has 0 days of day data.


 73%|███████████████████████████▋          | 1893/2599 [00:09<00:03, 225.45it/s]

RPTX only has 70 days of day data.
RSI only has 84 days of day data.
RKT only has 37 days of day data.
QS only has 17 days of day data.


 75%|████████████████████████████▎         | 1938/2599 [00:09<00:03, 200.06it/s]

ZI only has 81 days of day data.
RLAY only has 52 days of day data.


 77%|█████████████████████████████▏        | 2000/2599 [00:10<00:03, 180.88it/s]

HRMY only has 28 days of day data.


 79%|█████████████████████████████▊        | 2042/2599 [00:10<00:02, 188.30it/s]

YALA only has 0 days of day data.
WMG only has 82 days of day data.
MEG only has 47 days of day data.


 81%|██████████████████████████████▊       | 2104/2599 [00:10<00:02, 197.87it/s]

VERX only has 43 days of day data.
U only has 7 days of day data.
VNT only has 0 days of day data.
VITL only has 41 days of day data.


 82%|███████████████████████████████▎      | 2144/2599 [00:10<00:02, 181.36it/s]

UWMC only has 4 days of day data.


 85%|████████████████████████████████▏     | 2199/2599 [00:11<00:02, 173.14it/s]

CRSR only has 4 days of day data.
PCVX only has 75 days of day data.


 87%|█████████████████████████████████▏    | 2273/2599 [00:11<00:01, 176.25it/s]

THRY only has 0 days of day data.
NKTX only has 56 days of day data.


 91%|██████████████████████████████████▌   | 2368/2599 [00:12<00:01, 182.53it/s]

TALK only has 40 days of day data.


 93%|███████████████████████████████████▍  | 2427/2599 [00:12<00:00, 190.55it/s]

HPK only has 22 days of day data.
MAXN only has 22 days of day data.


 98%|█████████████████████████████████████▎| 2555/2599 [00:13<00:00, 206.08it/s]

ACI only has 65 days of day data.
SII only has 66 days of day data.


100%|██████████████████████████████████████| 2599/2599 [00:13<00:00, 194.69it/s]


In [52]:
# Checking out how many tickers are missing day data since 6/1/20
len(missing_days_tickers)

106

Indeed, we see that even requiring just day data from 6/1/20 to 9/30/20, we're missing 106 days of tickers. Furthermore, we're unsure as to how clean that data really is. Instead, we'll get the day data ourselves from the FRD original 1-min data.

## Checking the original FRD 1-min data to see if there's enough days of past data
Before we use the original FRD 1-min data, let's check if we have the appropriate number of days for the 1-min data. If there isn't, we might have to pivot and not use those extra days.

In [66]:
# Getting the AAPL path to check for appropriate number of days
AAPL_path = '/Volumes/T7/Original_FRD_Data_Till_Dec_2024/unzipped-FRD-csv-full/AAPL_full_1min_UNADJUSTED.csv'
AAPL_df = pd.read_csv(AAPL_path)
AAPL_df['Datetime'] = pd.to_datetime(AAPL_df['Datetime'])
start_date = pd.to_datetime("2020-06-01")
end_date = pd.to_datetime("2020-10-01")

# Getting the number of days from 6/1/20-9/30/20 and from 10/1/20 to the end of the data
num_days_extra = AAPL_df[(AAPL_df['Datetime'] >= start_date) & (AAPL_df['Datetime'] < end_date)]['Datetime'].dt.date.nunique()
num_days_orig = AAPL_df[(AAPL_df['Datetime'] >= end_date)]['Datetime'].dt.date.nunique()
print("Number of days of data from 6/1/20 to 9/30/20:", num_days_extra)
print("Number of days of data since 10/1/20 (original):", num_days_orig)

Number of days of data from 6/1/20 to 9/30/20: 86
Number of days of data since 10/1/20 (original): 1055


Firstly, we confirm that we still have 1,055 days (as expected), and we also confirm again that we have 86 days from 6/1/20-9/30/20.

In [67]:
# Setting some initial paths and variables
root_path_orig = '/Volumes/T7/Original_FRD_Data_Till_Dec_2024/unzipped-FRD-csv-full'
missing_days_tickers = []

# Cycling through all the appropriate tickers to make sure we have the necessary data
for ticker in tqdm(ticker_list_filtered):

    # Getting the ticker file name and path
    ticker_1min_file = ticker + "_full_1min_UNADJUSTED.csv"
    ticker_1min_path = os.path.join(root_path_orig, ticker_1min_file)

    # Reading in the csv and converting to Datetime column to proper datetime format
    df = pd.read_csv(ticker_1min_path)
    df['Datetime'] = pd.to_datetime(df['Datetime'])

    # Checking to see if the number of days is not equal to the expected number (86 days from 6/1/20-9/30/20)
    num_days_ticker = df[(df['Datetime'] >= start_date) & (df['Datetime'] < end_date)]['Datetime'].dt.date.nunique()
    if num_days_ticker != num_days_extra:
        print(ticker, "only has", num_days_ticker, "days of 1-min data from 6/1/20-9/30/20.")
        missing_days_tickers.append(ticker)


  0%|▏                                        | 10/2599 [00:03<10:57,  3.94it/s]

ASLE only has 69 days of 1-min data from 6/1/20-9/30/20.


  1%|▎                                        | 17/2599 [00:07<20:14,  2.13it/s]

ASAN only has 1 days of 1-min data from 6/1/20-9/30/20.


  1%|▌                                        | 38/2599 [00:19<14:00,  3.05it/s]

ACCD only has 63 days of 1-min data from 6/1/20-9/30/20.


  3%|█▏                                       | 72/2599 [00:46<21:34,  1.95it/s]

AOUT only has 28 days of 1-min data from 6/1/20-9/30/20.


  3%|█▏                                       | 73/2599 [00:46<18:26,  2.28it/s]

API only has 67 days of 1-min data from 6/1/20-9/30/20.


  3%|█▍                                       | 88/2599 [00:54<19:04,  2.19it/s]

ANNX only has 48 days of 1-min data from 6/1/20-9/30/20.


  4%|█▋                                      | 109/2599 [01:11<24:31,  1.69it/s]

ALIT only has 53 days of 1-min data from 6/1/20-9/30/20.


  5%|█▉                                      | 128/2599 [01:25<18:18,  2.25it/s]

AMWL only has 10 days of 1-min data from 6/1/20-9/30/20.


  6%|██▏                                     | 145/2599 [01:38<23:38,  1.73it/s]

ALXO only has 53 days of 1-min data from 6/1/20-9/30/20.


  6%|██▍                                     | 155/2599 [01:46<24:29,  1.66it/s]

ADV only has 84 days of 1-min data from 6/1/20-9/30/20.


  6%|██▍                                     | 159/2599 [01:47<11:29,  3.54it/s]

CMPS only has 9 days of 1-min data from 6/1/20-9/30/20.


  7%|██▋                                     | 176/2599 [02:03<20:26,  1.98it/s]

AZEK only has 77 days of 1-min data from 6/1/20-9/30/20.


  7%|██▊                                     | 179/2599 [02:04<11:36,  3.47it/s]

DOGZ only has 85 days of 1-min data from 6/1/20-9/30/20.


  8%|███▏                                    | 207/2599 [02:26<33:48,  1.18it/s]

AEVA only has 74 days of 1-min data from 6/1/20-9/30/20.


 11%|████▌                                   | 295/2599 [03:46<32:40,  1.18it/s]

CHPT only has 78 days of 1-min data from 6/1/20-9/30/20.


 12%|████▊                                   | 315/2599 [04:04<21:22,  1.78it/s]

CELU only has 82 days of 1-min data from 6/1/20-9/30/20.


 13%|█████▍                                  | 350/2599 [04:29<25:17,  1.48it/s]

BUR only has 74 days of 1-min data from 6/1/20-9/30/20.


 15%|█████▉                                  | 383/2599 [04:50<23:18,  1.58it/s]

IMTX only has 63 days of 1-min data from 6/1/20-9/30/20.


 17%|██████▋                                 | 435/2599 [05:29<22:56,  1.57it/s]

DYN only has 10 days of 1-min data from 6/1/20-9/30/20.


 18%|███████▏                                | 470/2599 [05:54<18:33,  1.91it/s]

AVO only has 0 days of 1-min data from 6/1/20-9/30/20.


 18%|███████▎                                | 477/2599 [05:58<17:31,  2.02it/s]

AVPT only has 2 days of 1-min data from 6/1/20-9/30/20.


 18%|███████▎                                | 478/2599 [05:59<14:44,  2.40it/s]

BSY only has 6 days of 1-min data from 6/1/20-9/30/20.


 20%|████████                                | 523/2599 [06:32<19:36,  1.76it/s]

DNB only has 64 days of 1-min data from 6/1/20-9/30/20.


 20%|████████                                | 525/2599 [06:34<22:18,  1.55it/s]

DM only has 85 days of 1-min data from 6/1/20-9/30/20.


 21%|████████▎                               | 544/2599 [06:44<13:35,  2.52it/s]

FFIE only has 19 days of 1-min data from 6/1/20-9/30/20.


 21%|████████▍                               | 549/2599 [06:48<17:13,  1.98it/s]

BNL only has 10 days of 1-min data from 6/1/20-9/30/20.


 27%|██████████▋                             | 695/2599 [08:49<25:09,  1.26it/s]

DADA only has 82 days of 1-min data from 6/1/20-9/30/20.


 28%|███████████▏                            | 727/2599 [09:11<12:58,  2.40it/s]

FROG only has 11 days of 1-min data from 6/1/20-9/30/20.


 29%|███████████▍                            | 741/2599 [09:21<21:40,  1.43it/s]

BIGC only has 40 days of 1-min data from 6/1/20-9/30/20.


 29%|███████████▋                            | 759/2599 [09:35<18:01,  1.70it/s]

HIMS only has 84 days of 1-min data from 6/1/20-9/30/20.


 29%|███████████▊                            | 764/2599 [09:37<12:00,  2.55it/s]

GBIO only has 77 days of 1-min data from 6/1/20-9/30/20.


 32%|████████████▋                           | 822/2599 [10:22<16:48,  1.76it/s]

GLSI only has 4 days of 1-min data from 6/1/20-9/30/20.


 32%|████████████▋                           | 827/2599 [10:25<14:32,  2.03it/s]

GOCO only has 55 days of 1-min data from 6/1/20-9/30/20.


 33%|█████████████▎                          | 866/2599 [10:56<12:18,  2.35it/s]

LPRO only has 78 days of 1-min data from 6/1/20-9/30/20.


 34%|█████████████▌                          | 881/2599 [11:09<14:37,  1.96it/s]

FOUR only has 82 days of 1-min data from 6/1/20-9/30/20.


 35%|█████████████▉                          | 903/2599 [11:21<10:15,  2.75it/s]

MP only has 63 days of 1-min data from 6/1/20-9/30/20.


 35%|█████████████▉                          | 905/2599 [11:23<15:13,  1.85it/s]

ML only has 33 days of 1-min data from 6/1/20-9/30/20.


 36%|██████████████▍                         | 941/2599 [11:53<12:02,  2.29it/s]

GCMG only has 59 days of 1-min data from 6/1/20-9/30/20.


 37%|██████████████▊                         | 964/2599 [12:08<11:55,  2.29it/s]

GDRX only has 6 days of 1-min data from 6/1/20-9/30/20.


 39%|███████████████                        | 1007/2599 [12:40<13:39,  1.94it/s]

ETWO only has 75 days of 1-min data from 6/1/20-9/30/20.


 40%|███████████████▍                       | 1027/2599 [12:58<16:49,  1.56it/s]

PTVE only has 10 days of 1-min data from 6/1/20-9/30/20.


 41%|███████████████▊                       | 1055/2599 [13:18<12:38,  2.03it/s]

CVAC only has 33 days of 1-min data from 6/1/20-9/30/20.


 41%|███████████████▉                       | 1058/2599 [13:20<11:25,  2.25it/s]

LUNG only has 0 days of 1-min data from 6/1/20-9/30/20.


 41%|███████████████▉                       | 1064/2599 [13:25<19:09,  1.34it/s]

LSPD only has 14 days of 1-min data from 6/1/20-9/30/20.


 41%|████████████████                       | 1073/2599 [13:30<12:28,  2.04it/s]

LMND only has 63 days of 1-min data from 6/1/20-9/30/20.


 41%|████████████████                       | 1074/2599 [13:30<11:53,  2.14it/s]

LI only has 44 days of 1-min data from 6/1/20-9/30/20.


 43%|████████████████▌                      | 1105/2599 [13:53<12:46,  1.95it/s]

JAMF only has 50 days of 1-min data from 6/1/20-9/30/20.


 43%|████████████████▋                      | 1109/2599 [13:57<21:02,  1.18it/s]

LEGN only has 82 days of 1-min data from 6/1/20-9/30/20.


 43%|████████████████▋                      | 1110/2599 [13:58<18:23,  1.35it/s]

LCID only has 9 days of 1-min data from 6/1/20-9/30/20.


 43%|████████████████▊                      | 1122/2599 [14:05<13:32,  1.82it/s]

LAZR only has 27 days of 1-min data from 6/1/20-9/30/20.


 44%|████████████████▉                      | 1131/2599 [14:13<20:18,  1.21it/s]

KYMR only has 28 days of 1-min data from 6/1/20-9/30/20.


 44%|█████████████████▏                     | 1142/2599 [14:18<11:41,  2.08it/s]

BEPC only has 48 days of 1-min data from 6/1/20-9/30/20.


 45%|█████████████████▌                     | 1169/2599 [14:35<13:02,  1.83it/s]

MIR only has 29 days of 1-min data from 6/1/20-9/30/20.


 45%|█████████████████▋                     | 1175/2599 [14:39<12:41,  1.87it/s]

CURI only has 73 days of 1-min data from 6/1/20-9/30/20.


 47%|██████████████████▍                    | 1229/2599 [15:17<10:41,  2.14it/s]

PRLD only has 4 days of 1-min data from 6/1/20-9/30/20.


 47%|██████████████████▍                    | 1231/2599 [15:17<07:18,  3.12it/s]

PLBY only has 21 days of 1-min data from 6/1/20-9/30/20.


 48%|██████████████████▌                    | 1236/2599 [15:21<10:46,  2.11it/s]

ORGN only has 18 days of 1-min data from 6/1/20-9/30/20.


 48%|██████████████████▊                    | 1256/2599 [15:38<09:40,  2.31it/s]

JBI only has 85 days of 1-min data from 6/1/20-9/30/20.


 49%|██████████████████▉                    | 1266/2599 [15:45<14:41,  1.51it/s]

ITOS only has 48 days of 1-min data from 6/1/20-9/30/20.


 50%|███████████████████▍                   | 1293/2599 [16:06<09:31,  2.29it/s]

PRCH only has 83 days of 1-min data from 6/1/20-9/30/20.


 50%|███████████████████▌                   | 1301/2599 [16:12<10:38,  2.03it/s]

PMVP only has 4 days of 1-min data from 6/1/20-9/30/20.


 50%|███████████████████▋                   | 1312/2599 [16:22<17:18,  1.24it/s]

PLRX only has 84 days of 1-min data from 6/1/20-9/30/20.


 51%|███████████████████▊                   | 1323/2599 [16:31<14:43,  1.44it/s]

PLTR only has 1 days of 1-min data from 6/1/20-9/30/20.


 52%|████████████████████                   | 1340/2599 [16:44<13:43,  1.53it/s]

BEKE only has 34 days of 1-min data from 6/1/20-9/30/20.


 53%|████████████████████▊                  | 1388/2599 [17:17<10:21,  1.95it/s]

NNOX only has 28 days of 1-min data from 6/1/20-9/30/20.


 54%|█████████████████████                  | 1402/2599 [17:29<16:43,  1.19it/s]

OPEN only has 71 days of 1-min data from 6/1/20-9/30/20.


 56%|██████████████████████                 | 1467/2599 [18:21<07:59,  2.36it/s]

NTST only has 34 days of 1-min data from 6/1/20-9/30/20.


 57%|██████████████████████▏                | 1481/2599 [18:29<09:35,  1.94it/s]

NRIX only has 48 days of 1-min data from 6/1/20-9/30/20.


 58%|██████████████████████▌                | 1506/2599 [18:49<06:44,  2.70it/s]

STR only has 81 days of 1-min data from 6/1/20-9/30/20.


 60%|███████████████████████▏               | 1547/2599 [19:23<09:12,  1.90it/s]

STEP only has 11 days of 1-min data from 6/1/20-9/30/20.


 60%|███████████████████████▍               | 1560/2599 [19:33<08:53,  1.95it/s]

NCNO only has 56 days of 1-min data from 6/1/20-9/30/20.


 60%|███████████████████████▌               | 1569/2599 [19:39<10:04,  1.70it/s]

MVST only has 83 days of 1-min data from 6/1/20-9/30/20.


 63%|████████████████████████▍              | 1630/2599 [20:22<10:03,  1.61it/s]

SNOW only has 11 days of 1-min data from 6/1/20-9/30/20.


 63%|████████████████████████▌              | 1638/2599 [20:29<12:52,  1.24it/s]

EOSE only has 69 days of 1-min data from 6/1/20-9/30/20.


 65%|█████████████████████████▎             | 1687/2599 [21:08<04:41,  3.24it/s]

RBOT only has 18 days of 1-min data from 6/1/20-9/30/20.


 67%|██████████████████████████             | 1733/2599 [21:41<10:00,  1.44it/s]

RNA only has 77 days of 1-min data from 6/1/20-9/30/20.


 68%|██████████████████████████▌            | 1767/2599 [22:05<07:51,  1.76it/s]

RPRX only has 75 days of 1-min data from 6/1/20-9/30/20.


 69%|██████████████████████████▉            | 1796/2599 [22:28<11:17,  1.19it/s]

RXT only has 40 days of 1-min data from 6/1/20-9/30/20.


 70%|███████████████████████████▏           | 1808/2599 [22:36<06:17,  2.10it/s]

XPEV only has 24 days of 1-min data from 6/1/20-9/30/20.


 71%|███████████████████████████▊           | 1851/2599 [23:08<05:30,  2.26it/s]

RPTX only has 72 days of 1-min data from 6/1/20-9/30/20.


 71%|███████████████████████████▊           | 1852/2599 [23:08<04:46,  2.61it/s]

RSI only has 84 days of 1-min data from 6/1/20-9/30/20.


 72%|████████████████████████████           | 1868/2599 [23:18<05:18,  2.30it/s]

RKT only has 39 days of 1-min data from 6/1/20-9/30/20.


 72%|████████████████████████████▏          | 1877/2599 [23:22<05:41,  2.11it/s]

QS only has 31 days of 1-min data from 6/1/20-9/30/20.


 73%|████████████████████████████▌          | 1906/2599 [23:43<06:50,  1.69it/s]

ZI only has 83 days of 1-min data from 6/1/20-9/30/20.


 74%|████████████████████████████▋          | 1915/2599 [23:49<05:44,  1.98it/s]

RLAY only has 54 days of 1-min data from 6/1/20-9/30/20.


 76%|█████████████████████████████▌         | 1971/2599 [24:27<03:35,  2.91it/s]

HRMY only has 30 days of 1-min data from 6/1/20-9/30/20.


 77%|██████████████████████████████▏        | 2010/2599 [25:02<04:49,  2.04it/s]

YALA only has 1 days of 1-min data from 6/1/20-9/30/20.


 78%|██████████████████████████████▍        | 2029/2599 [25:15<04:01,  2.36it/s]

WMG only has 84 days of 1-min data from 6/1/20-9/30/20.


 78%|██████████████████████████████▌        | 2033/2599 [25:17<04:15,  2.22it/s]

MEG only has 49 days of 1-min data from 6/1/20-9/30/20.


 80%|███████████████████████████████        | 2072/2599 [25:49<07:31,  1.17it/s]

VERX only has 45 days of 1-min data from 6/1/20-9/30/20.


 80%|███████████████████████████████▏       | 2075/2599 [25:49<04:02,  2.16it/s]

U only has 9 days of 1-min data from 6/1/20-9/30/20.


 81%|███████████████████████████████▍       | 2097/2599 [26:02<04:17,  1.95it/s]

VNT only has 3 days of 1-min data from 6/1/20-9/30/20.


 81%|███████████████████████████████▋       | 2108/2599 [26:08<03:42,  2.21it/s]

VITL only has 43 days of 1-min data from 6/1/20-9/30/20.


 81%|███████████████████████████████▊       | 2116/2599 [26:13<04:26,  1.81it/s]

UWMC only has 6 days of 1-min data from 6/1/20-9/30/20.


 84%|████████████████████████████████▋      | 2181/2599 [27:02<03:50,  1.81it/s]

CRSR only has 6 days of 1-min data from 6/1/20-9/30/20.


 84%|████████████████████████████████▉      | 2191/2599 [27:09<03:59,  1.71it/s]

PCVX only has 77 days of 1-min data from 6/1/20-9/30/20.


 86%|█████████████████████████████████▋     | 2243/2599 [27:51<04:30,  1.32it/s]

THRY only has 0 days of 1-min data from 6/1/20-9/30/20.


 87%|█████████████████████████████████▉     | 2264/2599 [28:08<02:06,  2.64it/s]

NKTX only has 58 days of 1-min data from 6/1/20-9/30/20.


 90%|███████████████████████████████████▏   | 2345/2599 [29:11<01:54,  2.22it/s]

TALK only has 42 days of 1-min data from 6/1/20-9/30/20.


 92%|████████████████████████████████████   | 2402/2599 [29:54<02:04,  1.58it/s]

HPK only has 24 days of 1-min data from 6/1/20-9/30/20.


 93%|████████████████████████████████████▏  | 2413/2599 [30:03<01:33,  1.99it/s]

MAXN only has 24 days of 1-min data from 6/1/20-9/30/20.


 97%|█████████████████████████████████████▊ | 2524/2599 [31:29<00:41,  1.80it/s]

ACI only has 67 days of 1-min data from 6/1/20-9/30/20.


 98%|██████████████████████████████████████▎| 2553/2599 [31:55<00:42,  1.07it/s]

SII only has 66 days of 1-min data from 6/1/20-9/30/20.


100%|███████████████████████████████████████| 2599/2599 [32:35<00:00,  1.33it/s]


In [68]:
# Checking to see how many days of missing data we have:
len(missing_days_tickers)

103

We now see that we still have a substantial amount of missing data, likely because the ticker didn't exist/was still private until 10/1/20, or that it simply didn't have the data. In either case, I've decided to stick with the data since 10/1/20 and perform my analyses from that. It's unfortunate that we'll be missing the extra days since 6/1/20, but it might be for the better anyway since that might introduce existing noise from the COVID-19 crash.

## Obtaining the cleaned 1-day data from the cleaned FRD 1-min dataset
Because I still don't trust the day data from FRD, I'll need to create it myself from the FRD 1-min data (which I've already cleaned anyway).

In [84]:
# Getting the source and destination directories and defining the RTH time window
input_root = "/Volumes/T7/Filtered-Cleaned-Parquet-ML"
output_root = "/Volumes/T7/Filtered-Cleaned-Parquet-1day"
os.makedirs(output_root, exist_ok=True)
rth_start = pd.to_datetime("09:30", format="%H:%M").time()
rth_end = pd.to_datetime("15:59", format="%H:%M").time()

# Cycling through each folder to get the 1-day parquet files
for folder_name in tqdm(os.listdir(input_root)):
    if not folder_name.endswith("_1min_parquet"):
        continue
    
    ticker = folder_name.split("_")[0]
    folder_path = os.path.join(input_root, folder_name)

    try:
        # Reading all intraday parquet files into one DataFrame and adding helper columns
        df = pd.read_parquet(folder_path)
        df['Date'] = df.index.date
        df['Time'] = df.index.time

        # Defining the daily rows that we'll ultimately add to
        daily_rows = []

        # Grouping by trading day
        for date, day_df in df.groupby('Date'):
            rth_df = day_df[(day_df['Time'] >= rth_start) & (day_df['Time'] <= rth_end)]
            pm_df = day_df[day_df['Time'] < rth_start]

            if rth_df.empty:  # Skip if no valid RTH data
                print(date, "for", ticker, "is missing data.")
                continue

            row = {
                "Datetime": pd.Timestamp(date),
                "Open": rth_df.iloc[0]['Open'],
                "High": rth_df['High'].max(),
                "Low": rth_df['Low'].min(),
                "Close": rth_df.iloc[-1]['Close'],
                "Volume": rth_df['Volume'].sum(),
                "PM_High": pm_df['High'].max() if not pm_df.empty else None,
                "PM_Low": pm_df['Low'].min() if not pm_df.empty else None,
                "PM_Volume": pm_df['Volume'].sum() if not pm_df.empty else None,
                "Ticker": ticker
            }

            daily_rows.append(row)

        # Saving daily data
        if daily_rows:
            df_out = pd.DataFrame(daily_rows)
            df_out = df_out.set_index("Datetime")
            output_path = os.path.join(output_root, f"{ticker}_1day.parquet")
            df_out.to_parquet(output_path)

    except Exception as e:
        print(f"Failed processing {folder_name}: {e}")

100%|███████████████████████████████████████| 2599/2599 [49:52<00:00,  1.15s/it]


In [85]:
# We can check to make sure that we still have the same number of files as before
len(os.listdir("/Volumes/T7/Filtered-Cleaned-Parquet-1day"))

2599

In [86]:
# We can also check to make sure the files look appropriate
AAPL_1day_path = '/Volumes/T7/Filtered-Cleaned-Parquet-1day/AAPL_1day.parquet'
pd.read_parquet(AAPL_1day_path)

Open    High     Low   Close       Volume   PM_High   PM_Low  \
Datetime                                                                      
2020-10-01  117.700  117.72  115.83  116.80   92076075.0  118.1000  116.510   
2020-10-02  112.840  115.37  112.22  113.02  116084401.0  117.0800  112.560   
2020-10-05  113.920  116.65  113.55  116.54   84421170.0  114.7100  113.300   
2020-10-06  115.680  116.12  112.25  113.16  132805466.0  116.7900  115.230   
2020-10-07  114.640  115.55  114.13  115.05   78303638.0  114.7500  113.134   
...             ...     ...     ...     ...          ...       ...      ...   
2024-12-04  242.870  244.11  241.25  243.02   20217543.0  243.6900  242.044   
2024-12-05  243.990  244.54  242.13  243.08   18743262.0  244.2767  242.100   
2024-12-06  242.905  244.63  242.08  242.86   19207219.0  243.2300  242.410   
2024-12-09  241.900  247.24  241.75  246.82   24141189.0  243.0000  241.570   
2024-12-10  246.910  248.21  245.34  247.80   19346976.0  248.0000  246.134   

            PM_Volume Ticker  
Datetime                      
2020-10-01  1755406.0   AAPL  
2020-10-02  4460967.0   AAPL  
2020-10-05  1690948.0   AAPL  
2020-10-06  1358666.0   AAPL  
2020-10-07  2071097.0   AAPL  
...               ...    ...  
2024-12-04   466815.0   AAPL  
2024-12-05   206871.0   AAPL  
2024-12-06   169685.0   AAPL  
2024-12-09   193561.0   AAPL  
2024-12-10   286748.0   AAPL  

[1055 rows x 9 columns]

We now have cleaned day data that we can use for every ticker, and we can now finally start our machine learning process. I will perform my analyses in a separate notebook, using all the cleaned data from my SSD.

## (Update) Adding in PM Highs, Lows, Close, and total Volume up till 9:27am and 9:28am
I thought of this after the fact, but I realized that if I solely want to use PM metrics up till either 9:27am or 9:28am and didn't want to pull out the minute data right away, I'll need pre-calculated metrics for them. The reasoning is simply because I want to potentially trade at the Open, which means that my model will need to run before the Open occurs. Using a PM_threshold of 9:28am will give me one minute for inference, whereas a PM_threshold of 9:27am will give me two minutes. I'm not sure exactly how long I'll need, so I'll get both metrics to be safe.

The PM High, Low, and Volume metrics will be calculated the same way as before, but instead of going to 9:29am, they will only go to 9:27am or 9:28am. The PM_Close at either time will be the Close price of that CS, and this will act as the new "Open" price when calculating metrics such as daily gap ups and such.

In [39]:
# Getting the different paths and defining the PM time cutoffs
minute_root = "/Volumes/T7/Filtered-Cleaned-Parquet-ML"
day_root = "/Volumes/T7/Filtered-Cleaned-Parquet-1day"
pm_cutoff_0927 = pd.to_datetime("09:27", format="%H:%M").time()
pm_cutoff_0928 = pd.to_datetime("09:28", format="%H:%M").time()

# Processing each ticker's 1-day parquet file
for file_name in tqdm(os.listdir(day_root)):
    if not file_name.endswith("_1day.parquet"):
        continue

    ticker = file_name.split("_")[0]
    day_file_path = os.path.join(day_root, file_name)
    minute_folder_path = os.path.join(minute_root, f"{ticker}_1min_parquet")

    try:
        # Loading the 1-day summary df
        day_df = pd.read_parquet(day_file_path)
        day_df.index = pd.to_datetime(day_df.index)

        # Loading all the 1-min data for this ticker
        df_minute = pd.read_parquet(minute_folder_path)
        df_minute.index = pd.to_datetime(df_minute.index)

        # Adding in Date and Time columns for grouping and filtering
        df_minute['Date'] = df_minute.index.date
        df_minute['Time'] = df_minute.index.time

        # Preparing lists for new columns
        pm_high_0927_list = []
        pm_low_0927_list = []
        pm_close_0927_list = []
        pm_volume_0927_list = []

        pm_high_0928_list = []
        pm_low_0928_list = []
        pm_close_0928_list = []
        pm_volume_0928_list = []

        # Grouping by date and calculating metrics
        for date, day_min_df in df_minute.groupby('Date'):
            if pd.Timestamp(date) not in day_df.index:
                # Skip dates not in the 1-day summary
                continue

            # Filtering for PM metrics up to 9:27am
            pm_0927_df = day_min_df[day_min_df['Time'] <= pm_cutoff_0927]
            if not pm_0927_df.empty:
                pm_high_0927_list.append(pm_0927_df['High'].max())
                pm_low_0927_list.append(pm_0927_df['Low'].min())
                pm_close_0927_list.append(pm_0927_df.iloc[-1]['Close'])
                pm_volume_0927_list.append(pm_0927_df['Volume'].sum())
            else:
                pm_high_0927_list.append(None)
                pm_low_0927_list.append(None)
                pm_close_0927_list.append(None)
                pm_volume_0927_list.append(None)
            
            # Filtering for PM metrics up to 9:28am
            pm_0928_df = day_min_df[day_min_df['Time'] <= pm_cutoff_0928]
            if not pm_0928_df.empty:
                pm_high_0928_list.append(pm_0928_df['High'].max())
                pm_low_0928_list.append(pm_0928_df['Low'].min())
                pm_close_0928_list.append(pm_0928_df.iloc[-1]['Close'])
                pm_volume_0928_list.append(pm_0928_df['Volume'].sum())
            else:
                pm_high_0928_list.append(None)
                pm_low_0928_list.append(None)
                pm_close_0928_list.append(None)
                pm_volume_0928_list.append(None)

        # Ensuring the lists align with day_df
        day_df = day_df.loc[pd.to_datetime([pd.Timestamp(d).date() for d in df_minute['Date'].unique()])]

        day_df['PM_High_0927'] = pm_high_0927_list
        day_df['PM_Low_0927'] = pm_low_0927_list
        day_df['PM_Close_0927'] = pm_close_0927_list
        day_df['PM_Volume_0927'] = pm_volume_0927_list

        day_df['PM_High_0928'] = pm_high_0928_list
        day_df['PM_Low_0928'] = pm_low_0928_list
        day_df['PM_Close_0928'] = pm_close_0928_list
        day_df['PM_Volume_0928'] = pm_volume_0928_list

        # Saving updated 1-day parquet file
        day_df.to_parquet(day_file_path)

    except Exception as e:
        print(f"Failed processing {file_name}: {e}")

100%|███████████████████████████████████████| 2599/2599 [37:18<00:00,  1.16it/s]


In [40]:
# We can check to make sure that we still have the same number of files as before
len(os.listdir("/Volumes/T7/Filtered-Cleaned-Parquet-1day"))

2599

In [41]:
# We can also check to make sure the files look appropriate
AAPL_1day_path = '/Volumes/T7/Filtered-Cleaned-Parquet-1day/AAPL_1day.parquet'
pd.read_parquet(AAPL_1day_path)

Open    High     Low   Close       Volume   PM_High   PM_Low  \
2020-10-01  117.700  117.72  115.83  116.80   92076075.0  118.1000  116.510   
2020-10-02  112.840  115.37  112.22  113.02  116084401.0  117.0800  112.560   
2020-10-05  113.920  116.65  113.55  116.54   84421170.0  114.7100  113.300   
2020-10-06  115.680  116.12  112.25  113.16  132805466.0  116.7900  115.230   
2020-10-07  114.640  115.55  114.13  115.05   78303638.0  114.7500  113.134   
...             ...     ...     ...     ...          ...       ...      ...   
2024-12-04  242.870  244.11  241.25  243.02   20217543.0  243.6900  242.044   
2024-12-05  243.990  244.54  242.13  243.08   18743262.0  244.2767  242.100   
2024-12-06  242.905  244.63  242.08  242.86   19207219.0  243.2300  242.410   
2024-12-09  241.900  247.24  241.75  246.82   24141189.0  243.0000  241.570   
2024-12-10  246.910  248.21  245.34  247.80   19346976.0  248.0000  246.134   

            PM_Volume Ticker  PM_High_0927  PM_Low_0927  PM_Close_0927  \
2020-10-01  1755406.0   AAPL        118.10      116.510       117.4500   
2020-10-02  4460967.0   AAPL        117.08      112.700       113.0000   
2020-10-05  1690948.0   AAPL        114.71      113.300       113.9700   
2020-10-06  1358666.0   AAPL        116.79      115.230       116.0100   
2020-10-07  2071097.0   AAPL        114.75      113.134       114.7000   
...               ...    ...           ...          ...            ...   
2024-12-04   466815.0   AAPL        243.69      242.044       242.8200   
2024-12-05   206871.0   AAPL        244.06      242.100       244.0600   
2024-12-06   169685.0   AAPL        243.20      242.410       242.9799   
2024-12-09   193561.0   AAPL        243.00      241.970       242.1200   
2024-12-10   286748.0   AAPL        248.00      246.134       247.2100   

            PM_Volume_0927  PM_High_0928  PM_Low_0928  PM_Close_0928  \
2020-10-01       1690772.0      118.1000      116.510       117.6800   
2020-10-02       4319384.0      117.0800      112.560       112.7500   
2020-10-05       1671728.0      114.7100      113.300       113.9200   
2020-10-06       1303941.0      116.7900      115.230       115.8500   
2020-10-07       2032261.0      114.7500      113.134       114.6000   
...                    ...           ...          ...            ...   
2024-12-04        457759.0      243.6900      242.044       243.0500   
2024-12-05        166440.0      244.2767      242.100       243.9400   
2024-12-06        163273.0      243.2300      242.410       242.9690   
2024-12-09        166778.0      243.0000      241.570       241.8899   
2024-12-10        282328.0      248.0000      246.134       247.0900   

            PM_Volume_0928  
2020-10-01       1726947.0  
2020-10-02       4389769.0  
2020-10-05       1685852.0  
2020-10-06       1331333.0  
2020-10-07       2042679.0  
...                    ...  
2024-12-04        463699.0  
2024-12-05        196704.0  
2024-12-06        168492.0  
2024-12-09        187306.0  
2024-12-10        284190.0  

[1055 rows x 17 columns]

With this updated data, we can now use PM prices that more closely reflect what will happen in reality. Furthermore, even though the process is slightly different from before (we're reading in the parquet files rather than creating them outright), it still took less time than before, which gives me more confidence that this Macbook can run more quickly than my previous Mac.